### PR TITLE
Es6 membership engagement banner

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

✨  ES6-ifies membership-engagement-banner 
✅  Reinstates tests that were removed in #17282 due to intermittent failures
🗑  Removes tests that were testing the behaviour of other modules such as [Message](https://github.com/guardian/frontend/blob/b25232007922899cec5a3ca70bdcf3c989512b2b/static/test/javascripts-legacy/spec/common/commercial/membership-engagement-banner.spec.js#L222-L238) and [membership-engagement-banner-tests](https://github.com/guardian/frontend/blob/b25232007922899cec5a3ca70bdcf3c989512b2b/static/test/javascripts-legacy/spec/common/commercial/membership-engagement-banner.spec.js#L177-L220) (?)
:atom: Simplifies tests by mocking all dependencies. The mediator mock provides a convenient synchronous API
🐛  Fixes a bug in which a typo was prevent interactions from ever being recorded 

## What is the value of this and can you measure success?

Fewer intermittent test failures, safer code

## Does this affect other platforms - Amp, Apps, etc?

No

## Tested in CODE?

No

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        pinOnHide: false,
+        siteMessageLinkName: 'membership message',
+        siteMessageCloseBtn: 'hide',
+        siteMessageComponentName: params.campaignCode,
+        trackDisplay: true,
+        cssModifierClass: colourClass,
+    }).show(renderedBanner);
+
+    if (messageShown) {
+        recordInteraction(params.interactionOnMessageShow);
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: This was previously passing `params.interactionOnMessageShown`, however [the relevant property](https://github.com/guardian/frontend/blob/master/static/src/javascripts-legacy/projects/common/modules/commercial/membership-engagement-banner-parameters.js#L25-L28) in `membership-engagement-banner-parameters` is `interactionOnMessageShow`

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    return `${prefix + campaignId}_${variantId}`;
+};
+
+const getUserVariantParams = (
+    userVariant: ?Variant,
+    campaignId: ?string,
+    defaultOffering: string
+): Object => {
+    if (campaignId && userVariant && userVariant.engagementBannerParams) {
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: I'm checking for the existence of `campaignId` here to avoid building a meaningless campaign code

**Comment by SiAdcock** [comment]
So little of the logic in this module is tested. I could probably test a few more side effects to improve the coverage, for instance ensuring a `new Message` was created with the correct parameters, or that `ophan.record` was called.

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.spec.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        emit(eventName, params) {
+            if (events[eventName]) {
+                events[eventName](params);
+            }
+        },
+        removeAllListeners() {
+            events = {};
+        },
+    };
+});
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.spec.js]: If this simplified synchronous mediator looks useful, we could probably move it to `lib/__mocks__`. I found that providing a synchronous API for the mock means we no longer have to emit events as test hooks.

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        : params.messageText;
+
+    const urlParameters = {
+        REFPVID: params.pageviewId,
+        INTCMP: params.campaignCode,
+    };
+
+    const linkUrl = `${params.linkUrl}?${constructQuery(urlParameters)}`;
+
+    const renderedBanner = template(messageTemplate, {
```
Comment (GHaberis) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Could we use es6 template literals here and drop the lodash `templates` library?

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    return Object.assign(
+        {},
+        defaultParams,
+        getUserVariantParams(userVariant, campaignId, defaultParams.offering)
+    );
+};
+
+// Used to send an interaction if the engagement banner is shown.
+const recordInteraction = (interaction: Object): void => {
+    if (interaction) {
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: According to the type definition, this check should not be necessary. Which is the right one? :)

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        const component = interaction.component;
+        const value = interaction.value;
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: `const { component,  value} = interaction` ?

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        if (component && value) {
+            ophan.record({
+                component,
+                value,
+            });
+        }
+    }
+};
+
+const paypalAndCreditCardImage =
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Could you move this to the top of the file since it's only a static assignment? Could also be a good usage of

```js
config.get('images.acquisitions.paypal-and-credit-card', '')
```

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    (config.images &&
+        config.images.acquisitions &&
+        config.images.acquisitions['paypal-and-credit-card']) ||
+    '';
+
+const selectSequentiallyFrom = (array: Array<any>): any =>
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Any chance of making the type definition for array more precise?

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    array[local.get('gu.alreadyVisited') % array.length];
+
+const showBanner = (params: Object): void => {
+    if (params === DO_NOT_RENDER_ENGAGEMENT_BANNER || isBlocked()) {
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Wait: if this `=== DO_NOT_RENDER_ENGAGEMENT_BANNER` could be ever true, the type of `params` should allow string.

Don't know about `isBlocked` but sounds like this is a boolean.

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    const messageShown = new Message(messageCode, {
+        pinOnHide: false,
+        siteMessageLinkName: 'membership message',
+        siteMessageCloseBtn: 'hide',
+        siteMessageComponentName: params.campaignCode,
+        trackDisplay: true,
+        cssModifierClass: colourClass,
+    }).show(renderedBanner);
+
+    if (messageShown) {
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: How could this ever be falsy?

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        recordInteraction(params.interactionOnMessageShow);
+
+        mediator.emit('membership-message:display');
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Did you check if something is listening for this? Found so many of them in the past which aren't used anymore ...

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 10 above)
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: `Message.prototype.show` [can return false](https://github.com/guardian/frontend/blob/master/static/src/javascripts-legacy/projects/common/modules/ui/message.js#L60) under certain circumstances

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 11 above)
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: This is being listened for in [`membership-engagement-banner-tests`](https://github.com/guardian/frontend/blob/master/static/src/javascripts-legacy/projects/common/modules/experiments/tests/membership-engagement-banner-tests.js#L43)

**Comment by SiAdcock** [comment]
Behold! My masterstroke!

@GHaberis @sndrs @gustavpursche please could you give it another peruse

**Comment by dominickendrick** [comment]
Hey, Great work Simon! I would just run this by someone on @guardian/contributions just so we can check there is no financial impact when this is released. This component makes us a lot of money

**Comment by SiAdcock** [comment]
No probs, @dominickendrick! Happy to discuss 😊

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    }
+};
+
+const membershipEngagementBannerInit = (): Promise<void> =>
+    get().then(location => {
+        const bannerParams = deriveBannerParams(location);
+
+        if (
+            bannerParams &&
+            (local.get('gu.alreadyVisited') || 0) >= bannerParams.minArticles
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Since `'gu.alreadyVisited'` is used in two places, would it be better to store it in a variable at the top? Or have a wrapper, which returns the localStorage entry?

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+import mediator from 'lib/mediator';
+import { testCanBeRun } from 'common/modules/experiments/test-can-run-checks';
+import MembershipEngagementBannerTests from 'common/modules/experiments/tests/membership-engagement-banner-tests';
+import { inlineSvg } from 'common/views/svgs';
+import { isInTest, variantFor } from 'common/modules/experiments/segment-util';
+import { epicEngagementBannerTests } from 'common/modules/experiments/acquisition-test-selector';
+import membershipEngagementBannerUtils from 'common/modules/commercial/membership-engagement-banner-parameters';
+import { isBlocked } from 'common/modules/commercial/membership-engagement-banner-block';
+import ophan from 'ophan/ng';
+import { get } from 'lib/geolocation';
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: The named import makes a bit unhappy. What about `import { get as getGeoLocation } from 'lib/geolocation'`?

**Review by gu-stav — APPROVED** [review]
Only minor comments. Very nice work overall!

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 13 above)
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Good shout. Would be nice to fix this everywhere it's used, will do this in a separate PR

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+import { constructQuery } from 'lib/url';
+
+type Interaction = {
+    component: 'string',
+    value: 'string',
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: lol good catch!

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 12 above)
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: I'll extract this into a function. I'm disinclined to save this as a variable at the top as I'm working to remove side effects.

**Review by joelochlann — APPROVED** [review]
Great work, thanks @SiAdcock! Just a few small Flow-related points

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+};
+
+// Used to send an interaction if the engagement banner is shown.
+// #? A lot of null checking here, but it's hard to reason about the possible values
+//    `interaction` could take. Perhaps we can simplify when we have more Flow coverage
+const recordInteraction = (interaction: ?Interaction): void => {
+    if (interaction) {
+        const { component, value } = interaction;
+
+        if (component && value) {
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: I was being overly cautious by not removing these checks, as it is difficult to reason about this code. However, I think you're right, this check is unnecessary

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 2 above)
Comment (jranks123) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Nice!

**Review by jranks123 — COMMENTED** [review]
👍  Looks good, let's just keep a close eye on banner conversions after it goes out

**Comment by SiAdcock** [comment]
Thanks for the comments, @joelochlann, @jranks123. I've made a few changes as a result. Adding an `EngagementBannerParams` type really makes things clearer!

I'll merge this tomorrow morning as it's a bit late in the day today. I'll drop you a line when this is done so we can keep an eye on engagements. I'll also give it one more test on CODE, just to make sure.
